node/core/runtime-api/src/cache.rs

```diff
@@ -0,0 +1,301 @@
+	dmq_contents: MemoryLruCache<(Hash, ParaId), DmqContents>,
+	inbound_hrmp_channels_contents: MemoryLruCache<(Hash, ParaId), InboundHrmpChannelsContents>,
+}
+
+struct Validators(Vec<ValidatorId>);
+struct ValidatorGroups((Vec<Vec<ValidatorIndex>>, GroupRotationInfo));
+struct AvailabilityCores(Vec<CoreState>);
+struct PPersistedValidationData(Option<PersistedValidationData>);
+struct CheckValidationOutputs(bool);
+struct SessionIndexForChild(SessionIndex);
+struct VValidationCode(Option<ValidationCode>);
+struct HistoricalValidationCode(Option<ValidationCode>);
+struct CandidatePendingAvailability(Option<CommittedCandidateReceipt>);
+struct CandidateEvents(Vec<CandidateEvent>);
+struct SSessionInfo(Option<SessionInfo>);
+struct DmqContents(Vec<InboundDownwardMessage<BlockNumber>>);
+struct InboundHrmpChannelsContents(BTreeMap<ParaId, Vec<InboundHrmpMessage<BlockNumber>>>);
+
+impl ResidentSize for Validators {
+	fn resident_size(&self) -> usize {
+		self.0.len() * 32
+	}
+}
+
+impl ResidentSize for ValidatorGroups {
+	fn resident_size(&self) -> usize {
+		let len = self.0.0.iter().fold(0, |acc, x| acc + x.len() * 4);
+		std::mem::size_of::<GroupRotationInfo>() + len
+	}
+}
+
+impl ResidentSize for AvailabilityCores {
+	fn resident_size(&self) -> usize {
+		self.0.len() * std::mem::size_of::<CoreState>()
+	}
+}
+
+impl ResidentSize for PPersistedValidationData {
+	fn resident_size(&self) -> usize {
+		std::mem::size_of::<PersistedValidationData>()
+	}
+}
+
+impl ResidentSize for CheckValidationOutputs {
+	fn resident_size(&self) -> usize {
+		std::mem::size_of::<SessionIndex>()
+	}
+}
+
+impl ResidentSize for SessionIndexForChild {
+	fn resident_size(&self) -> usize {
+		4
+	}
+}
+
+impl ResidentSize for VValidationCode {
+	fn resident_size(&self) -> usize {
+		match self.0 {
+			Some(ref code) => code.0.len(),
+			None => 1,
+		}
+	}
+}
+impl ResidentSize for HistoricalValidationCode {
+	fn resident_size(&self) -> usize {
+		match self.0 {
+			Some(ref code) => code.0.len(),
+			None => 32,
+		}
+	}
+}
+
+impl ResidentSize for CandidatePendingAvailability {
+	fn resident_size(&self) -> usize {
+		std::mem::size_of::<CandidatePendingAvailability>()
+	}
+}
+
+impl ResidentSize for CandidateEvents {
+	fn resident_size(&self) -> usize {
+		self.0.len() * std::mem::size_of::<CandidateEvent>()
+	}
+}
+
+impl ResidentSize for SSessionInfo {
+	fn resident_size(&self) -> usize {
+		std::mem::size_of::<SessionInfo>()
+	}
+}
+
+impl ResidentSize for DmqContents {
+	fn resident_size(&self) -> usize {
+		self.0.len() * std::mem::size_of::<InboundDownwardMessage<BlockNumber>>()
+	}
+}
+
+impl ResidentSize for InboundHrmpChannelsContents {
+	fn resident_size(&self) -> usize {
+		let values_size = self.0.iter().fold(0, |acc, x| acc + x.1.len()) *
+			std::mem::size_of::<InboundHrmpMessage<BlockNumber>>();
+		let keys_size = self.0.len() * std::mem::size_of::<ParaId>();
+
+		values_size + keys_size
+	}
+}
```

Comment:
once we change the layout of `InboundHrmpChannelsContents`, this might be invalidated and there is no test to catch it
this is the reason we have `MallocSizeOf` from `parity-util-mem`

Reply:
Added, could you take a look? I am not sure I'm using it correctly in all cases